src/view/subgraph.rs

```diff
@@ -0,0 +1,347 @@
+//! Views into non-hierarchical parts of `PortView`s.
+
+use std::collections::{BTreeSet, HashMap};
+
+use crate::{algorithms::ConvexChecker, Direction, LinkView, NodeIndex, PortIndex, PortView};
+
+use super::filter::PortFiltered;
+
+type NodeCallback = fn(NodeIndex, &SubgraphContext) -> bool;
+type PortCallback = fn(PortIndex, &SubgraphContext) -> bool;
+
+/// View into a subgraph of a PortView.
+///
+/// The subgraph is given by boundary edges that define where one "enters" the
+/// subgraph (incoming boundary edge) and one "leaves" it (outgoing boundary
+/// edges).
+///
+/// A subgraph is well-defined if the incoming/outgoing boundary edges partition
+/// the edges between the children of the root node into three sets:
+///  - boundary edges: either incoming boundary or outgoing boundary edges,
+///  - interior edges: edges such that all the successor edges are either
+///    outgoing boundary edges or interior edges AND all the predecessor edges
+///    are either incoming boundary edges or interior edges,
+///  - exterior edges: edges such that all the successor edges are either
+///    incoming boundary edges or exterior edges AND all the predecessor edges
+///    are either outgoing boundary edges or exterior edges.
+///
+/// Then the subgraph is made of the interior edges and contains all nodes that
+/// are
+///  - adjacent to an interior edge, or
+///  - in-adjacent to an incoming edge and out-adjacent to an outgoing edge.
```

Comment:
Shouldn't this be the inverse?
I don't think we use "in-adjacent" anywhere else, so maybe use "source/target" instead?